src/util/settings.cpp

```diff
@@ -0,0 +1,161 @@
+            result = reverse_precedence ? source.span.begin()[0] : source.span.end()[-1];
+        } else if (source.span.last_negated()) {
+            result = false;
+        }
+    });
+    return result;
+}
+
+std::vector<SettingsValue> GetSettingsList(const Settings& settings,
+    const std::string& section,
+    const std::string& name,
+    bool ignore_default_section_config)
+{
+    std::vector<SettingsValue> result;
+    bool result_complete = false;
+    bool prev_negated_empty = false;
+    MergeSettings(settings, section, name, [&](Source source) {
+        // Weird behavior preserved for backwards compatibility: Apply config
+        // file settings even if negated on command line. Negating a setting on
+        // command line will discard earlier settings on the command line and
+        // settings in the config file, unless the negated command line value is
+        // followed by non-negated value, in which case config file settings
+        // will be brought back from the dead (but earlier command line settings
+        // will still be discarded).
+        const bool add_zombie_config_values = source.config_file && !prev_negated_empty;
+
+        // Ignore settings in top-level config section if requested.
+        if (ignore_default_section_config && source.config_file_top_level) return;
+
+        // Add new settings to the result if isn't already complete, or if the
+        // values are zombies.
+        if (!result_complete || add_zombie_config_values) {
+            for (const auto& value : source.span) {
+                if (value.isArray()) {
+                    result.insert(result.end(), value.getValues().begin(), value.getValues().end());
+                } else {
+                    result.push_back(value);
+                }
+            }
+        }
+
+        // If a setting was negated, or if a setting was forced, set
+        // result_complete to true to ignore any later lower priority settings.
+        result_complete |= source.span.negated() > 0 || source.forced;
+
+        // Update the negated and empty state used for the zombie values check.
+        prev_negated_empty |= source.span.last_negated() && result.empty();
+    });
+    return result;
+}
+
+bool HasIgnoredDefaultSectionConfigValue(const Settings& settings, const std::string& section, const std::string& name)
+{
+    bool has_ignored = true;
+    MergeSettings(settings, section, name, [&](Source source) {
+        // If top-level config value is unset, or any other value is set,
+        // then no top-level setting is being ignored.
+        has_ignored &= source.config_file_top_level ? !source.span.empty() : source.span.empty();
+    });
+    return has_ignored;
+}
+
+SettingsSpan::SettingsSpan(const std::vector<SettingsValue>& vec) : SettingsSpan(vec.data(), vec.size()) {}
+const SettingsValue* SettingsSpan::begin() const { return data + negated(); }
+const SettingsValue* SettingsSpan::end() const { return data + size; }
```

Comment:
`end()` could be `noexcept`, not sure about all these other trivial functions (with regards to `isFalse()` via `negated()`).